server/form-pages/apply/risks-and-needs/risk-of-serious-harm/custom-forms/oasysImport.ts

```diff
@@ -1,25 +1,162 @@
 
   body: OasysImportBody
 
+  taskData: string
+
+  hasOasysRecord: boolean
+
+  oasysCompleted: string
+
+  oasysStarted: string
+
+  noOasysBannerText = `No OASys record available to import for ${this.personName}`
+
+  noOasysDescriptiveText = `No information can be imported for the risk of serious harm section because ${this.personName} 
+                            does not have a Layer 3 OASys completed in the last 6 months.`
+
+  taskName = 'risk-of-serious-harm'
+
   constructor(
     body: Partial<OasysImportBody>,
     private readonly application: Application,
+    oasys: OASysRiskOfSeriousHarm,
+    taskData: string,
   ) {
     this.body = body as OasysImportBody
+    this.hasOasysRecord = (oasys && Boolean(Object.keys(oasys).length)) || false
+    if (oasys) {
+      this.oasysStarted = oasys.dateStarted && DateFormats.isoDateToUIDate(oasys.dateStarted, { format: 'medium' })
+      this.oasysCompleted =
+        oasys.dateCompleted && DateFormats.isoDateToUIDate(oasys.dateCompleted, { format: 'medium' })
+    }
+    this.taskData = taskData
+  }
+
+  static async initialize(
+    body: Partial<OasysImportBody>,
+    application: Application,
+    token: string,
+    dataServices: DataServices,
+  ) {
+    let oasys
+    let taskDataJson
+
+    if (!application.data['risk-of-serious-harm']) {
+      try {
+        oasys = await dataServices.personService.getOasysRoSH(token, application.person.crn)
+
+        taskDataJson = JSON.stringify(OasysImport.getTaskData(oasys))
+      } catch (e) {
+        if (e.status === 404) {
+          oasys = null
+        } else {
+          throw e
+        }
+      }
+      return new OasysImport(body, application, oasys, taskDataJson)
+    }
+    if (OasysImport.isRoSHApplicationDataImportedFromOASys(application)) {
+      return new Summary(application.data['risk-of-serious-harm'], application)
+    }
+    return new RiskToOthers(application.data['risk-of-serious-harm'], application)
+  }
+
+  private static isRoSHApplicationDataImportedFromOASys(application: Application): boolean {
+    const rosh = application.data['risk-of-serious-harm']
+    if (
+      (rosh?.['risk-to-others'] && 'dateOfOasysImport' in rosh['risk-to-others']) ||
+      (rosh?.['risk-factors'] && 'dateOfOasysImport' in rosh['risk-factors']) ||
+      (rosh?.['reducing-risk'] && 'dateOfOasysImport' in rosh['reducing-risk'])
+    ) {
+      return true
+    }
+    return false
+  }
+
+  private static getTaskData(oasysSections: OASysRiskOfSeriousHarm): Partial<RoSHTaskData> {
+    const taskData = { 'risk-of-serious-harm': {} } as Partial<RoSHTaskData>
+    const today = new Date()
+
+    oasysSections.rosh.forEach(question => {
+      switch (question.questionNumber) {
+        case 'R10.1':
+          taskData['risk-of-serious-harm']['risk-to-others'] = {
+            ...taskData['risk-of-serious-harm']['risk-to-others'],
```

Comment:
why do we spread the data here? there shouldn't be any existing data under `risk-of-serious-harm` right?

Reply:
because we're storing multiple question answers under the 'risk-to-others' and 'risk-factors' keys, so I wanted to avoid overwriting e.g. the 'whoIsAtRisk' answer when we try to write the 'natureOfRisk' answer